src/Acquisition/_Acquisition.c

```diff
@@ -949,28 +966,58 @@ Wrapper_call(Wrapper *self, PyObject *args, PyObject *kw)
 }
 
 /* Code to handle accessing Wrapper objects as sequence objects */
-
 static Py_ssize_t
-Wrapper_length(Wrapper *self)
-{
-  long l;
-  PyObject *r;
+Wrapper_length(PyObject* self)
+{
+    PyObject *result;
+    PyObject *callable;
+    PyObject *tres;
+    Py_ssize_t res;
+
+    callable = PyObject_GetAttr(self, py__len__);
+    if (callable == NULL) {
+        if (PyErr_ExceptionMatches(PyExc_AttributeError)) {
+            /* PyObject_LengthHint in Python3 catches only TypeError.
+             * Python2 catches both (type and attribute error)
+             */
+            PyErr_SetString(PyExc_TypeError, "object has no len()");
+        }
+        return -1;
+    }
 
-  UNLESS(r=PyObject_GetAttr(OBJECT(self), py__len__)) return -1;
-  UNLESS_ASSIGN(r,PyObject_CallObject(r,NULL)) return -1;
-  l=PyInt_AsLong(r);
-  Py_DECREF(r);
-  return l;
+    result = PyObject_CallObject(callable, NULL);
+    Py_DECREF(callable);
+
+    if (result == NULL) {
+        return -1;
+    }
+
+    /* PyLong_AsSsize_t can only be called on long objects. */
+    tres = PyNumber_Long(result);
+    Py_DECREF(result);
+
+    if (tres == NULL) {
+        return -1;
+    }
+
+    res = PyLong_AsSsize_t(tres);
+    Py_DECREF(tres);
+
+    if (res == -1 && PyErr_Occurred()) {
+        return -1;
+    }
+
+    return res;
 }
 
 static PyObject *
-Wrapper_add(Wrapper *self, PyObject *bb)
+Wrapper_add(PyObject *self, PyObject *bb)
 {
-  return CallMethodO(OBJECT(self),py__add__,Build("(O)", bb) ,NULL);
+  return CallMethodO(self, py__add__,Build("(O)", bb) ,NULL);
 }
 
```

Comment:
Indeed this could be replaced by the marco.
Apart from that I see more potential in simplifying _Acquisition.c.
I'll try to do some during the weekend. However I'm going to push them to a separate branch, I don't like to pollute the current branch. Since these changes are independent of the python3 port.